apps/glusterfs/app_volume.go

```diff
@@ -328,3 +328,49 @@ func (a *App) VolumeExpand(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+func (a *App) VolumeClone(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	vol_id := vars["id"]
+
+	var msg api.VolumeCloneRequest
+	err := utils.GetJsonFromRequest(r, &msg)
+	if err != nil {
+		http.Error(w, "request unable to be parsed", http.StatusUnprocessableEntity)
+		return
+	}
+	err = msg.Validate()
+	if err != nil {
+		http.Error(w, "validation failed: "+err.Error(),
+			http.StatusBadRequest)
+		logger.LogError("validation failed: " + err.Error())
+		return
+	}
+
+	var volume *VolumeEntry
+	err = a.db.View(func(tx *bolt.Tx) error {
+		var err error
+		volume, err = NewVolumeEntryFromId(tx, vol_id)
+		if err == ErrNotFound {
```

Comment:
I think we need a check here for volume visibility. See VolumeInfo at around line 195 in app_volume.go.
This will protect us from accidentally starting a clone operation on a pending volume and will protect us from stuff like a clone coming in on a volume that's being deleted.

Reply:
Ah, got it!